core/src/sumeragi/network_topology.rs

```diff
@@ -234,4 +356,202 @@ mod tests {
         topology.update_peer_list(new_peers);
         assert_eq!(extract_ports(&topology), vec![0, 2, 5, 7])
     }
+
+    #[test]
+    fn filter_by_role() {
```

Comment:
Very good unit test. I'd recommend that you make it parametric, so that it accepts the number of key pairs (the macro would have to be inlined), so we could convert this to a `proptest`.